specifications/FiniteMonotonic/Constrain_CRDT.tla

```diff
@@ -0,0 +1,55 @@
+-------------------------- MODULE Constrain_CRDT ----------------------------
+
+EXTENDS Naturals
+
+CONSTANT Node
+
+VARIABLES counter, converge
```

Comment:
> Really our convergence property will only become true if all requests stop for a while. We model this by adding a boolean variable, converge, which is arbitrarily set to TRUE at some point and stops additional increments from occurring. Since this is a virtual variable that wouldn’t exist in any real system, we should move it out into a model-checking spec that imports our “good copy” spec and wraps its definitions with converge logic:

Adding `converge` changes the safety part of `Spec`, i.e., `Init /\ [][Next]_v`.  In TLA, we don't have to change `Init /\ [][Next]_v` but can conjoin a suitable fairness constraint.  A suitable constraint, though not machine-closed, would be `[]<>S!Convergence`.

Reply:
I'm not entirely sure I follow, and this might be related to your suggestion of a liveness property up above of `\A n, o \in Node: []<>(counter[n] = counter[o])`. I don't think that property would actually hold even if we could model-check an infinite-state-space spec, right? I can imagine a behavior where some node keeps incrementing before a gossip reaches it.

Reply:
Hence the fairness constraint.

Reply:
Does TLC support that? I thought it only supported fairness constraints. That's pretty neat! Do we lose anything by discarding machine closure?

Reply:
I thought we call any liveness formula that we conjoin to `Init /\ [][Next]_v` a fairness constraint!?  For example, I call the liveness formulas `[]<>S!Convergence`, `WF_v(Next)`, `\A n, o \in Node: WF_v(Gossip(n,o))`, ... fairness constraints when conjoined to `Init /\ [][Next]_v`.

In CRDT, `[]<>S!Convergence` works fine as a fairness constraint with TLC. It is also easy to see that `Init /\ [][Next]_v /\ []<>S!Convergence => []<>S!Convergence` (assuming `Liveness` is defined as `[]<>S!Convergence`). `Init /\ [][Next]_v /\ \A n, o \in Node: WF_v(Gossip(n,o)) => []<>S!Convergence` is less easy to see but should also hold.  However, that fairness constraint is [machine closed](https://groups.google.com/g/tlaplus/c/-L1mCJZ6-BA/m/Vd-Cp_-XAwAJ).  I would argue that we don't care about machine closure in CRDT.

Reply:
Fairness constraint is a good term for it, I really meant I thought TLC only supported WF and SF assertions in that place.

>`Init /\ [][Next]_v /\ \A n, o \in Node: WF_v(Gossip(n,o)) => []<>S!Convergence` is less easy to see but should also hold.

Would this still hold if GarbageCollect is included? It seems like there should be an increment > gossip > increment > gossip > garbagecollect loop that would keep it from being satisfied.

Reply:
> Would this still hold if GarbageCollect is included? It seems like there should be an increment > gossip > increment > gossip > garbagecollect loop that would keep it from being satisfied.

Apologies, I didn't pay enough attention to the definition of Gossip.  We have to strengthen the fairness constraint to e.g. (non-mc) `WF_vars(Next) /\ \A n, o \in Node : <>[][Gossip(n,o)]_vars`.  This constraint is stronger than it has to be.